File: Server/src/main/java/com/karandaev/roadCrackDetectionServer/image/decode/ImageDimensionProbe.java
package com.karandaev.roadCrackDetectionServer.image.decode;

import com.karandaev.roadCrackDetectionServer.image.exception.ImageValidationException;
import com.karandaev.roadCrackDetectionServer.image.config.ImageSecurityProperties;
import org.springframework.http.HttpStatus;

import javax.imageio.ImageIO;
import javax.imageio.ImageReader;
import javax.imageio.stream.ImageInputStream;
import java.awt.Dimension;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.Iterator;

/** Reads and validates image dimensions without fully decoding image pixels. */
public final class ImageDimensionProbe {

  private final ImageSecurityProperties props;

  /**
   * Creates an image dimension probe.
   *
   * @param props image security configuration
   */
  public ImageDimensionProbe(ImageSecurityProperties props) {
    this.props = props;
  }

  /**
   * Reads image dimensions using ImageIO readers without fully decoding pixel data.
   *
   * @param bytes raw image bytes
   * @return detected image dimensions
   * @throws ImageValidationException if dimensions cannot be read
   */
  public Dimension probeDimensions(byte[] bytes) {
    try (ImageInputStream inputStream =
        ImageIO.createImageInputStream(new ByteArrayInputStream(bytes))) {

      if (inputStream == null) {
        throw new ImageValidationException(
            HttpStatus.UNSUPPORTED_MEDIA_TYPE, "Cannot read image stream");
      }

      Iterator<ImageReader> readers = ImageIO.getImageReaders(inputStream);
      if (!readers.hasNext()) {
        throw new ImageValidationException(
            HttpStatus.UNSUPPORTED_MEDIA_TYPE, "No ImageReader for this format");
      }

      ImageReader reader = readers.next();
      try {
        reader.setInput(inputStream, true, true);

        return new Dimension(reader.getWidth(0), reader.getHeight(0));
      } finally {
        reader.dispose();
      }
    } catch (IOException error) {
      throw new ImageValidationException(
          HttpStatus.UNSUPPORTED_MEDIA_TYPE, "Corrupted image header");
    }
  }

  /**
   * Validates image dimensions against configured width, height, and megapixel limits.
   *
   * @param dimensions detected image dimensions
   * @throws ImageValidationException if dimensions are invalid or too large
   */
  public void validateDimensions(Dimension dimensions) {
    int width = dimensions.width;
    int height = dimensions.height;

    if (width <= 0 || height <= 0) {
      throw new ImageValidationException(
          HttpStatus.UNSUPPORTED_MEDIA_TYPE, "Invalid image dimensions");
    }

    if (width > props.getMaxWidth() || height > props.getMaxHeight()) {
      throw new ImageValidationException(
          HttpStatus.PAYLOAD_TOO_LARGE, "Image dimensions too large");
    }

    long pixels = (long) width * (long) height;
    long maxPixels = (long) props.getMaxMegapixels() * 1_000_000L;

    if (pixels > maxPixels) {
      throw new ImageValidationException(
          HttpStatus.PAYLOAD_TOO_LARGE, "Image megapixels too large");
    }
  }
}

File: MobileApp/app/src/main/java/com/karandaev/geo_inspect/core/image/validation/decode/ImageDimensionProbe.kt
package com.karandaev.geo_inspect.core.image.validation.decode

import android.graphics.BitmapFactory
import com.karandaev.geo_inspect.core.image.validation.config.ImageSecurityProperties
import com.karandaev.geo_inspect.core.image.validation.exception.ImageValidationErrorCode
import com.karandaev.geo_inspect.core.image.validation.exception.ImageValidationException

/**
 * Reads and validates image dimensions without fully decoding image pixels.
 */
class ImageDimensionProbe(
  private val props: ImageSecurityProperties
) {

  /**
   * Reads image dimensions without decoding the full bitmap into memory.
   *
   * @param bytes Raw image bytes.
   * @return Detected image dimensions.
   * @throws ImageValidationException if dimensions cannot be read.
   */
  fun probeDimensions(bytes: ByteArray): ImageDimensions {
    val options = BitmapFactory.Options().apply {
      inJustDecodeBounds = true
    }

    BitmapFactory.decodeByteArray(
      bytes,
      0,
      bytes.size,
      options
    )

    if (options.outWidth <= 0 || options.outHeight <= 0) {
      throw ImageValidationException(
        code = ImageValidationErrorCode.InvalidImageDimensions,
        message = "Invalid image dimensions."
      )
    }

    return ImageDimensions(
      width = options.outWidth,
      height = options.outHeight
    )
  }

  /**
   * Validates image dimensions against configured width, height, and megapixel limits.
   *
   * @param dimensions Detected image dimensions.
   * @throws ImageValidationException if dimensions are invalid or too large.
   */
  fun validateDimensions(dimensions: ImageDimensions) {
    val width = dimensions.width
    val height = dimensions.height

    if (width <= 0 || height <= 0) {
      throw ImageValidationException(
        code = ImageValidationErrorCode.InvalidImageDimensions,
        message = "Invalid image dimensions."
      )
    }

    if (width > props.maxWidth || height > props.maxHeight) {
      throw ImageValidationException(
        code = ImageValidationErrorCode.ImageDimensionsTooLarge,
        message = "Image dimensions are too large."
      )
    }

    val pixels = width.toLong() * height.toLong()
    val maxPixels = props.maxMegapixels.toLong() * PixelsInMegapixel

    if (pixels > maxPixels) {
      throw ImageValidationException(
        code = ImageValidationErrorCode.ImageMegapixelsTooLarge,
        message = "Image megapixels are too large."
      )
    }
  }

  private companion object {
    const val PixelsInMegapixel = 1_000_000L
  }
}